### PR TITLE
Add finetuning support for `Speaker` class

### PR DESCRIPTION
**Pull request by ain-soph**
To allow tuning for `speaker_embeds` tensor. The PR includes:

1. Rename: `_sample_random -> sample_random_tensor` to allow usage out of class.
2. The `spk_emb` argument in `Speaker.apply()` changes `str -> Union[str, torch.Tensor]`
3. Add `inplace` argument to support gradient backward when setting to `False`
4. Update return type to return `torch.Tensor`

Besides, I don't quite see the advantage and motivation to encode the `speaker_embeds` tensor to string. To me, it only makes things more complex. I would personally prefer cpu `torch.Tensor` or `np.ndarray`

cc @fumiama

### CODE REVIEW

File: ChatTTS/model/speaker.py [hunk 1]
```diff
@@ -115,7 +120,7 @@ def decode_prompt(prompt: str) -> torch.Tensor:
         return torch.from_numpy(p.astype(np.int32)).view(*shp)
 
     @torch.no_grad()
-    def _sample_random(self) -> torch.Tensor:
+    def sample_random_tensor(self) -> torch.Tensor:
```
Comment (fumiama) [inline ChatTTS/model/speaker.py:123]: I agree with your changes but as for this name, it'd be better to leave it unchanged.

**Review by fumiama — CHANGES_REQUESTED** [review]
> Besides, I don't quite see the advantage and motivation to encode the speaker_embeds tensor to string.

People can still use the tensor but the .pt file can easily contains virus. The str encoding is equal to the numpy array dump but is easy to copy&paste.

File: ChatTTS/model/speaker.py (hunk 1 above)
Comment (ain-soph) [reply ChatTTS/model/speaker.py:123]: According to [PEP 8](https://peps.python.org/pep-0008/#method-names-and-instance-variables), 
> Use one leading underscore only for non-public methods and instance variables.
In this case, we need to call this method out of class. Therefore, it's better not to start with a leading underscore.

Comment (fumiama) [reply ChatTTS/model/speaker.py:123]: I know your meaning but we want to mark this method as unrecommended. Peopole should always use the str version except for some special reasons.

**Comment by fumiama** [comment]
Also, you can save the str to a txt file (with the ext .pt if you like) and use it just like a .pt file if you like.

File: ChatTTS/model/speaker.py (hunk 1 above)
Comment (ain-soph) [reply ChatTTS/model/speaker.py:123]: Yes, that makes sense if there is no plan to add fine-tuning script.
I'm currently working on my [fine-tuning branch](https://github.com/ain-soph/ChatTTS/tree/dev) and it seems to be working to some extent. Will the main repo maintainer consider a PR for supporting fine-tuning?

**Comment by ain-soph** [comment]
> Also, you can save the str to a txt file (with the ext .pt if you like) and use it just like a .pt file if you like.

"use it just like a .pt file if you like" -> You mean we could directly read it via `torch.load()` instead of using `pybase16384` to encode/decode? I didn't know that.

**Comment by fumiama** [comment]
> > Also, you can save the str to a txt file (with the ext .pt if you like) and use it just like a .pt file if you like.
> 
> 
> 
> "use it just like a .pt file if you like" -> You mean we could directly read the tensor via `torch.load()` instead of using `pybase16384` to encode/decode? I didn't know that.

Just treat the str as a kind of binary.

File: ChatTTS/model/speaker.py (hunk 1 above)
Comment (fumiama) [reply ChatTTS/model/speaker.py:123]: It sounds great, we will be happy to accept your conteibution. As for the fine-tuning, since we will put the codes in the Chat class, it can contain such a internal api directly.

Comment (ain-soph) [reply ChatTTS/model/speaker.py:123]: > It sounds great, we will be happy to accept your contribution. As for the fine-tuning, since we will put the codes in the Chat class, it can contain such an internal api directly.

That's where I feel confused. If you want to call `Speaker._sample_random()` method in `Chat` class, then it's an out-of-class call and it means this method is public. If that's the case, it's better to not start with leading underscore if we want to obey the PEP 8 naming standard.

Comment (ain-soph) [reply ChatTTS/model/speaker.py:123]: I see cases for non-class methods ([example](https://github.com/pytorch/vision/blob/f5d9ecd5024d1c4f21a5620cf2cca6d07eabd469/torchvision/prototype/datasets/utils/_resource.py#L21)), but I don't know if it's still the case for class methods.

Comment (fumiama) [reply ChatTTS/model/speaker.py:123]: Well, in this case, as our main purpose is to prevent users from using this method and the Speaker class is a single instance in Chat, we can regard it as a kind of `static` or `non-class`.

**Review by fumiama — APPROVED** [review]
Thanks!
